docs/architecture/adr-009-modular-antehandler.md

```diff
@@ -0,0 +1,173 @@
+
+My preferred approach is to split the AnteHandler functionality into tightly scoped "micro-functions", while preserving the one-after-the-other ordering that would come from the per-module approach.
+
+We can then have a way to chain these micro-functions so that they run one after the other. Modules may define multiple ante micro-functions and then also provide a default per-module AnteHandler that implements a default, suggested order for these micro-functions. 
+
+Users can order the AnteHandlers easily by simply using the ModuleManager. The ModuleManager will take in a list of AnteHandlers and return a single AnteHandler that runs each AnteHandler in the order of the list provided. If the user is comfortable with the default ordering of each module, this is as simple as providing a list with each module's antehandler (exactly the same as BeginBlocker and EndBlocker).
+
+If however, users wish to change the order or add, modify, or delete ante micro-functions in anyway; they can always define their own ante micro-functions and add them explicitly to the list that gets passed into module manager.
+
+#### Default Workflow:
+
+##### SDK code:
+
+```golang
+func Chainer(order []AnteHandler) AnteHandler {
+    return func(ctx Context, tx Tx, simulate bool) (newCtx Context, res Result, abort bool) {
+        for _, ante := range order {
+            ctx, res, abort := ante(ctx, tx, simulate)
+            if abort {
+                return ctx, res, abort
+            }
+        }
+        return ctx, res, abort
+    }
+}
+```
+
+```golang
+func VerifySignatures(ctx Context, tx Tx, simulate bool) (newCtx Context, res Result, abort bool) {
+    // verify signatures
+    // Returns InvalidSignature Result and abort=true if sigs invalid
+    // Return OK result and abort=false if sigs are valid
+}
+
+func ValidateMemo(ctx Context, tx Tx, simulate bool) (newCtx Context, res Result, abort bool) {
+    // validate memo
+}
+
+AuthModuleAnteHandler := Chainer([]AnteHandler{VerifySignatures, ValidateMemo})
+```
+
+```golang
+func DeductFees(ctx Context, tx Tx, simulate bool) (newCtx Context, res Result, abort bool) {
+    // Deduct fees from tx
+    // Abort if insufficient funds in account to pay for fees
+}
+
+func CheckMempoolFees(ctx Context, tx Tx, simulate bool) (newCtx Context, res  Result, abort bool) {
+    // If CheckTx: Abort if the fees are less than the mempool's minFee parameter
+}
+
+DistrModuleAnteHandler := Chainer([]AnteHandler{CheckMempoolFees, DeductFees})
+```
+
+```golang
+type ModuleManager struct {
+    // other fields
+    AnteHandlerOrder []AnteHandler
+}
+
+func (mm ModuleManager) GetAnteHandler() AnteHandler {
+    retun Chainer(mm.AnteHandlerOrder)
+}
+```
+
+##### User Code:
+
+```golang
+moduleManager.SetAnteHandlerOrder([]AnteHandler(AuthModuleAnteHandler, DistrModuleAnteHandler))
```

Comment:
I know you mention weave as hard to reason - but here, the actual stack is split over multiple modules, making it quite hard to see all the pieces and reason about it.

At the least, the entire flow should be visible in one place, to reduce the error of forgetting or duplicating some check (or placing in the wrong order).